# <center> 3) Preparación de los Datos <center>

## Selección de Datos

In [1]:
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
from IPython.display import display, HTML
# !(pip install chainladder)
import chainladder as cl

print("pandas: " + pd.__version__)
print("numpy: " + np.__version__)
print("chainladder: " + cl.__version__)

pandas: 1.4.2
numpy: 1.22.0
chainladder: 0.8.18


In [2]:
data=pd.read_csv('wkcomp_pos.csv')

In [6]:
data.columns
# data.columns[4:11]

Index(['GRCODE', 'GRNAME', 'AccidentYear', 'DevelopmentYear', 'DevelopmentLag',
       'IncurLoss_D', 'CumPaidLoss_D', 'BulkLoss_D', 'EarnedPremDIR_D',
       'EarnedPremCeded_D', 'EarnedPremNet_D', 'Single', 'PostedReserve97_D'],
      dtype='object')

In [7]:
triangle = cl.Triangle(
    data,
    origin="AccidentYear",
    development="DevelopmentYear",
    columns=data.columns[4:11],
    index=["GRNAME"],
    cumulative=True,
)
triangle

Triangle Summary
Valuation:                                            2006-12
Grain:                                                   OYDY
Shape:                                       (132, 7, 19, 19)
Index:                                               [GRNAME]
Columns:    [DevelopmentLag, IncurLoss_D, CumPaidLoss_D, B...

In [8]:
triangle.index.head()

GRNAME
0                  Agway Ins Co
1             Alaska Nat Ins Co
2    Alaska Timber Ins Exchange
3           Allstate Ins Co Grp
4  American Contractors Ins Grp

In [11]:
print("Estructura del Chainladder triangle:", type(triangle.values))
print("Suma de los valores de la tabla de datos:", np.nansum(triangle.values))

Estructura del Chainladder triangle: <class 'numpy.ndarray'>
Suma de los valores de la tabla de datos: 768244598.0


Triángulo de pérdidas

In [18]:
triangle[[x for x in data.columns if("Loss" in x)]]

Triangle Summary
Valuation:                                   2006-12
Grain:                                          OYDY
Shape:                              (132, 3, 19, 19)
Index:                                      [GRNAME]
Columns:    [IncurLoss_D, CumPaidLoss_D, BulkLoss_D]

Frecuencias de la parte exploratoria para GRNAME, la cual nos mostraba los nombres de las 132 empresas en la base de datos.

In [27]:
freq=pd.DataFrame(data['GRNAME'].value_counts())
freq.index

Index(['Allstate Ins Co Grp', 'Bancinsure Inc', 'Harco Natl Ins Co',
       'Transportation Cas Ins Co', 'Beacon Mut Ins Co', 'Mapfre Reins Corp',
       'National American Ins Co', 'Midwest Family Mut Ins Co',
       'Associated Industries Ins Co', 'Lumbermens Underwriting Alliance',
       ...
       'Catholic Relief Ins Co Of Amer', 'Care West Ins Co', 'FFVA Mut Ins Co',
       'MO Employers Mut Ins Co', 'Martingale Natl Ins Co',
       'Hyundai Marine & Fire Ins Co Ltd', 'Agway Ins Co', 'Mada Ins Exchange',
       'Lumber Ins Cos', 'Tower Ins Co Of NY'],
      dtype='object', length=132)

In [31]:
triangle.loc[freq.index[0]].iloc[-1]['CumPaidLoss_D']

12        24        36        48        60        72        84        96        108       120  132  144  156  168  180  192  204  216  228
1988  70571.0  155905.0  220744.0  251595.0  274156.0  287676.0  298499.0  304873.0  321808.0  325322.0  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN
1989  66547.0  136447.0  179142.0  211343.0  231430.0  244750.0  254557.0  270059.0  273873.0  277574.0  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN
1990  52233.0  133370.0  178444.0  204442.0  222193.0  232940.0  253337.0  256788.0  261166.0  263000.0  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN
1991  59315.0  128051.0  169793.0  196685.0  213165.0  234676.0  239195.0  245499.0  247131.0  248319.0  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN
1992  39991.0   89873.0  114117.0  133003.0  154362.0  159496.0  164013.0  166212.0  167397.0  168844.0  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN
1993  19744.0   47229.0   61909.0   85099.0   87215.0   88602.0   89444.0   89899.0   90446.0   90686.0  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN
1994  20379.0   46773.0   88636.0   91077.0   92583.0   93346.0   93897.0   94165.0   94558.0   94730.0  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN
1995  18756.0   84712.0   87311.0   89200.0   90001.0   90247.0   90687.0   91068.0   91001.0   91161.0  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN
1996  42609.0   44916.0   46981.0   47899.0   48583.0   49109.0   49442.0   49073.0   49161.0   49255.0  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN
1997    691.0    2085.0    2795.0    2866.0    2905.0    2909.0    2908.0    2909.0    2909.0    2909.0  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN
1998      NaN       NaN       NaN       NaN       NaN       NaN       NaN       NaN       NaN       NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN
1999      NaN       NaN       NaN       NaN       NaN       NaN       NaN       NaN       NaN       NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN
2000      NaN       NaN       NaN       NaN       NaN       NaN       NaN       NaN       NaN       NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN
2001      NaN       NaN       NaN       NaN       NaN       NaN       NaN       NaN       NaN       NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN
2002      NaN       NaN       NaN       NaN       NaN       NaN       NaN       NaN       NaN       NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN
2003      NaN       NaN       NaN       NaN       NaN       NaN       NaN       NaN       NaN       NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN
2004      NaN       NaN       NaN       NaN       NaN       NaN       NaN       NaN       NaN       NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN
2005      NaN       NaN       NaN       NaN       NaN       NaN       NaN       NaN       NaN       NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN
2006      NaN       NaN       NaN       NaN       NaN       NaN       NaN       NaN       NaN       NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN

In [33]:
triangle["CaseIncurLoss"] = triangle["IncurLoss_D"] - triangle["BulkLoss_D"]
triangle["PaidToInc"] = triangle["CumPaidLoss_D"] / triangle["IncurLoss_D"]
triangle[["CaseIncurLoss", "PaidToInc"]]

Triangle Summary
Valuation:                     2006-12
Grain:                            OYDY
Shape:                (132, 2, 19, 19)
Index:                        [GRNAME]
Columns:    [CaseIncurLoss, PaidToInc]

In [34]:
triangle["CumPaidLoss_D"].sum() / triangle["EarnedPremNet_D"].sum()

12        24        36        48        60        72        84        96        108       120  132  144  156  168  180  192  204  216  228
1988  0.169002  0.377577  0.511551  0.589170  0.641199  0.670077  0.691697  0.707762  0.726853  0.734252  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN
1989  0.171152  0.380515  0.510029  0.592723  0.641889  0.673263  0.695181  0.718333  0.727896  0.736776  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN
1990  0.170250  0.402847  0.540943  0.621712  0.669555  0.699547  0.727737  0.741724  0.753801  0.761300  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN
1991  0.168254  0.384413  0.510169  0.585788  0.633060  0.669131  0.685162  0.700841  0.709746  0.716366  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN
1992  0.156652  0.356895  0.463579  0.535713  0.585474  0.606911  0.626156  0.639385  0.648478  0.656737  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN
1993  0.137911  0.299451  0.386339  0.448436  0.478351  0.500951  0.514849  0.524425  0.532535  0.538605  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN
1994  0.135294  0.289192  0.382980  0.429647  0.460100  0.479147  0.491811  0.502014  0.509397  0.516786  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN
1995  0.131396  0.293704  0.367651  0.415741  0.444896  0.462494  0.475378  0.483372  0.490237  0.497043  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN
1996  0.157595  0.304066  0.387472  0.436638  0.467298  0.488593  0.503505  0.512192  0.520638  0.526420  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN
1997  0.154052  0.331010  0.426129  0.483107  0.517449  0.539827  0.554047  0.565653  0.571992  0.578609  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN
1998       NaN       NaN       NaN       NaN       NaN       NaN       NaN       NaN       NaN       NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN
1999       NaN       NaN       NaN       NaN       NaN       NaN       NaN       NaN       NaN       NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN
2000       NaN       NaN       NaN       NaN       NaN       NaN       NaN       NaN       NaN       NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN
2001       NaN       NaN       NaN       NaN       NaN       NaN       NaN       NaN       NaN       NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN
2002       NaN       NaN       NaN       NaN       NaN       NaN       NaN       NaN       NaN       NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN
2003       NaN       NaN       NaN       NaN       NaN       NaN       NaN       NaN       NaN       NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN
2004       NaN       NaN       NaN       NaN       NaN       NaN       NaN       NaN       NaN       NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN
2005       NaN       NaN       NaN       NaN       NaN       NaN       NaN       NaN       NaN       NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN
2006       NaN       NaN       NaN       NaN       NaN       NaN       NaN       NaN       NaN       NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN

In [35]:
triangle.groupby("GRNAME").sum()

Triangle Summary
Valuation:                                            2006-12
Grain:                                                   OYDY
Shape:                                       (132, 9, 19, 19)
Index:                                               [GRNAME]
Columns:    [DevelopmentLag, IncurLoss_D, CumPaidLoss_D, B...

In [39]:
pd.DataFrame(np.unique(triangle["GRNAME"]))

0
0                           Agway Ins Co
1                      Alaska Nat Ins Co
2             Alaska Timber Ins Exchange
3                    Allstate Ins Co Grp
4           American Contractors Ins Grp
..                                   ...
127        Wisconsin American Mut Ins Co
128          Wisconsin Cnty Mut Ins Corp
129                    Workers Comp Exch
130  Yasuda Fire & Marine Ins Co Of Amer
131                Zurich Ins (Guam) Inc

[132 rows x 1 columns]

In [37]:
triangle.sum(axis="index").sum(axis="segments")

Triangle Summary
Valuation:                                            2006-12
Grain:                                                   OYDY
Shape:                                         (1, 9, 19, 19)
Index:                                               [GRNAME]
Columns:    [DevelopmentLag, IncurLoss_D, CumPaidLoss_D, B...

In [38]:
triangle.sum(axis=0).sum(axis=1)

12            24            36            48            60            72            84            96            108           120  132  144  156  168  180  192  204  216  228
1988  6.505334e+06  6.998133e+06  7.251426e+06  7.372627e+06  7.516601e+06  7.562091e+06  7.569563e+06  7.595274e+06  7.627302e+06  7.628929e+06  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN
1989  6.993589e+06  7.485209e+06  7.703016e+06  7.929058e+06  8.019964e+06  8.051545e+06  8.085541e+06  8.122084e+06  8.143175e+06  8.135914e+06  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN
1990  7.497213e+06  8.222881e+06  8.450718e+06  8.597213e+06  8.651657e+06  8.678904e+06  8.729039e+06  8.765898e+06  8.777143e+06  8.763675e+06  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN
1991  8.307773e+06  8.883966e+06  9.129062e+06  9.234930e+06  9.289198e+06  9.361264e+06  9.399476e+06  9.425834e+06  9.415632e+06  9.407692e+06  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN
1992  8.511972e+06  9.181979e+06  9.294409e+06  9.316729e+06  9.384021e+06  9.438707e+06  9.473071e+06  9.476289e+06  9.494934e+06  9.489522e+06  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN
1993  9.280994e+06  9.671786e+06  9.746181e+06  9.725504e+06  9.784804e+06  9.818267e+06  9.810062e+06  9.782090e+06  9.781425e+06  9.771582e+06  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN
1994  9.631728e+06  9.962769e+06  9.981869e+06  9.963848e+06  9.984381e+06  9.994971e+06  9.983078e+06  9.979242e+06  9.985671e+06  1.000053e+07  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN
1995  9.607355e+06  9.916423e+06  9.953288e+06  9.866027e+06  9.890350e+06  9.871873e+06  9.881837e+06  9.885418e+06  9.915027e+06  9.945121e+06  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN
1996  8.998679e+06  9.221765e+06  9.366872e+06  9.354414e+06  9.347331e+06  9.366417e+06  9.382955e+06  9.404258e+06  9.439836e+06  9.467252e+06  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN
1997  8.271236e+06  8.668566e+06  8.855925e+06  8.865622e+06  8.899831e+06  8.956443e+06  8.984406e+06  9.026011e+06  9.066499e+06  9.083781e+06  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN
1998           NaN           NaN           NaN           NaN           NaN           NaN           NaN           NaN           NaN           NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN
1999           NaN           NaN           NaN           NaN           NaN           NaN           NaN           NaN           NaN           NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN
2000           NaN           NaN           NaN           NaN           NaN           NaN           NaN           NaN           NaN           NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN
2001           NaN           NaN           NaN           NaN           NaN           NaN           NaN           NaN           NaN           NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN
2002           NaN           NaN           NaN           NaN           NaN           NaN           NaN           NaN           NaN           NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN
2003           NaN           NaN           NaN           NaN           NaN           NaN           NaN           NaN           NaN           NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN
2004           NaN           NaN           NaN           NaN           NaN           NaN           NaN           NaN           NaN           NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN
2005           NaN           NaN           NaN           NaN           NaN           NaN           NaN           NaN           NaN           NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN
2006           NaN           NaN           NaN           NaN           NaN           NaN           NaN           NaN           NaN           NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN